### DISCUSSION TITLE
Erlang Message Hooks

### DISCUSSION THREAD

**Discussion by SSaiPrasad7** [emqx/emqx#4394] [Q&A] [answered]
Hello @zmstone. I hope you’re doing well.

1. I looked into the documentation as you provided earlier [here.](https://github.com/emqx/emqx/discussions/4333#discussioncomment-487228).As docs looks unclear to me because I am at beginning stage of Erlang as well as Emqx world.I request you to provide some more details about `how to setup with plugin (rebar3),connect to emqx and modify the code.`It would be great help if you provide me step by step procedure in order for building a plugin and connecting to emqx.
2. I have an another doubt about plugin named [ `emqx_plugin_template`](https://github.com/emqx/emqx-plugin-template).Please can you provide information about how to use these template and configure the file present in `/etc/emqx/plugins/` directory?Whether it is possible to create message hooks using this `emqx_plugin_template` ?

I researched a lot about emqx plugins and issues present on GitHub. Most issues are about emqx custom plugin configuration.It would be huge benefit for everyone.
Thanking you.

**Comment by zmstone** [comment, marked as answer]
Hi @saiprasadsamudrala 

Plugins are not developed separately, they need to be placed in the project and built as a part of it.

To develop/build/test a plugin, I believe [lib-extra/README.md](https://github.com/emqx/emqx/blob/master/lib-extra/README.md) is straightforward enough to get started.

If you build the project (latest master) with bash command `EMQX_EXTRA_PLUGINS=all make emqx-tar`
The built release in `_build/emqx/rel/emqx` is a self-contained tarball which you can use to deploy and run.
e.g. my build result is `_build/emqx/rel/emqx/emqx-4.3-beta.1-0501c97e.tar.gz`


As for configuration, you can put plugin config in plugin's `etc` directory,
the `.conf` files are collected by the build scripts and put in `etc/plugins` directory in the release.

**Comment by SSaiPrasad7** [comment]
Thank you @zmstone .I completed with Erlang hooks but I want to write my own print statements on console.
May I know location and name of the file where I can edit with my print statements?

**Reply by zmstone** [reply]
If it's for ad-hoc debugging purposes,
you can call `io:format(user, "~p~n", [TermToInspect])` from anywhere in your code to have it printed.

If you want it to be properly logged, you can call `emqx_logger` APIs

**Reply by SSaiPrasad7** [reply]
![hook_emqx](https://user-images.githubusercontent.com/36000620/112251804-e83e5100-8c81-11eb-9fe9-0666589e3626.png)
Hello @zmstone 
Please can u share information about from which file the particular highlighted message was printed?
(Location and Name of the file).

**Reply by zmstone** [reply]
Hi
It's printed from `emqx_tracer.erl`

**Reply by zmstone** [reply]
```
  trace(publish, #message{topic = <<"$SYS/", _/binary>>}) ->
      %% Do not trace '$SYS' publish
      ignore;
  trace(publish, #message{from = From, topic = Topic, payload = Payload})
          when is_binary(From); is_atom(From) ->
      emqx_logger:info(#{topic => Topic,
                         mfa => {?MODULE, ?FUNCTION_NAME, ?FUNCTION_ARITY} },
                       "PUBLISH to ~s: ~0p", [Topic, Payload]).```

**Reply by SSaiPrasad7** [reply]
Yeah got it @zmstone.
There is lot of debug information once loading the plugin.Can we display only `[info]` on the console?
How to comment out `[debug]` ?
Is it in `emqx_logger.erl`?

**Reply by zmstone** [reply]
you can set `log.level` to info in `etc/emqx.conf`
or override with environment variable `export EMQX_LOG__LEVEL=info`
